iceoryx_posh/test/integrationtests/test_popo_chunk_building_blocks.cpp

```diff
@@ -0,0 +1,190 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/internal/popo/building_blocks/chunk_distributor.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/chunk_receiver.hpp"
+#include "iceoryx_posh/internal/popo/building_blocks/chunk_sender.hpp"
+#include "iceoryx_posh/mepoo/mepoo_config.hpp"
+
+#include "test.hpp"
+
+#include <chrono>
+#include <stdlib.h>
+#include <thread>
+
+using namespace ::testing;
+using namespace iox::popo;
+using namespace iox::cxx;
+using namespace iox::mepoo;
+using namespace iox::posix;
+using ::testing::Return;
+
+struct DummySample
+{
+    uint64_t dummy{42};
+};
+
+static constexpr size_t MEMORY_SIZE = 1024 * 1024;
+uint8_t m_memory[MEMORY_SIZE];
+static constexpr uint32_t NUM_CHUNKS_IN_POOL = 3 * iox::MAX_RECEIVER_QUEUE_CAPACITY;
+static constexpr uint32_t ITERATIONS = 10000;
+static constexpr uint32_t SMALL_CHUNK = 128;
+static constexpr uint32_t MAX_NUMBER_QUEUES = 128;
+
+using ChunkDistributorData_t = iox::popo::ChunkDistributorData<MAX_NUMBER_QUEUES, iox::popo::ThreadSafePolicy>;
+using ChunkDistributor_t = iox::popo::ChunkDistributor<ChunkDistributorData_t>;
+
+class ChunkBuildingBlocks_IntegrationTest : public Test
+{
+  public:
+    ChunkBuildingBlocks_IntegrationTest()
+    {
+        m_mempoolConfig.addMemPool({SMALL_CHUNK, NUM_CHUNKS_IN_POOL});
+        m_memoryManager.configureMemoryManager(m_mempoolConfig, &m_memoryAllocator, &m_memoryAllocator);
+    }
+    virtual ~ChunkBuildingBlocks_IntegrationTest()
+    {
+        /// @note One chunk is on hold due to the fact that chunkSender and chunkDistributor hold last chunk
+        EXPECT_THAT(m_memoryManager.getMemPoolInfo(0).m_usedChunks, Eq(1));
+    }
+
+    void SetUp()
+    {
+        m_chunkSender.addQueue(&m_chunkQueueData);
+        m_chunkDistributor.addQueue(&m_chunkReceiverData);
+    }
+    void TearDown(){};
+
+    void publish()
+    {
+        for (size_t i = 0; i < ITERATIONS; i++)
+        {
+            m_chunkSender.allocate(sizeof(DummySample))
+                .on_success([&](iox::mepoo::ChunkHeader* chunkHeader) {
+                    auto sample = chunkHeader->payload();
+                    new (sample) DummySample();
+                    static_cast<DummySample*>(sample)->dummy = i;
+                    m_chunkSender.send(chunkHeader);
+                    m_sendCounter++;
+                })
+                .on_error([]() {
+                    // Errors shall never occur
+                    FAIL();
+                });
+
+            /// Add some jitter to make thread breathe
+            std::this_thread::sleep_for(std::chrono::nanoseconds(rand() % 100));
+        }
+    }
+
+    void forward()
+    {
+        uint64_t forwardCounter{0};
+        while (m_run)
```

Comment:
this could result in premature stop of the thread, even if there still are some elements in the queue. this could be prevented with a m_publisherRun and m_forwarderRun.

Reply:
Done